<?xml version="1.0" encoding="UTF-8"?><!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml" lang="en"><head><meta http-equiv="Content-Type" content="text/html;charset=UTF-8"/><link rel="stylesheet" href="../jacoco-resources/report.css" type="text/css"/><link rel="shortcut icon" href="../jacoco-resources/report.gif" type="image/gif"/><title>ObservableCreateTest.java</title><link rel="stylesheet" href="../jacoco-resources/prettify.css" type="text/css"/><script type="text/javascript" src="../jacoco-resources/prettify.js"></script></head><body onload="window['PR_TAB_WIDTH']=4;prettyPrint()"><div class="breadcrumb" id="breadcrumb"><span class="info"><a href="../jacoco-sessions.html" class="el_session">Sessions</a></span><a href="../index.html" class="el_report">RxJava_2_x$Unnamed.exec</a> &gt; <a href="index.source.html" class="el_package">io.reactivex.internal.operators.observable</a> &gt; <span class="el_source">ObservableCreateTest.java</span></div><h1>ObservableCreateTest.java</h1><pre class="source lang-java linenums">/**
 * Copyright (c) 2016-present, RxJava Contributors.
 *
 * Licensed under the Apache License, Version 2.0 (the &quot;License&quot;); you may not use this file except in
 * compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License is
 * distributed on an &quot;AS IS&quot; BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See
 * the License for the specific language governing permissions and limitations under the License.
 */

package io.reactivex.internal.operators.observable;

import static org.junit.Assert.*;

import java.io.IOException;
import java.util.List;

import org.junit.Test;

import io.reactivex.*;
import io.reactivex.disposables.*;
import io.reactivex.exceptions.TestException;
import io.reactivex.functions.Cancellable;
import io.reactivex.observers.TestObserver;
import io.reactivex.plugins.RxJavaPlugins;

<span class="fc" id="L30">public class ObservableCreateTest {</span>

    @Test(expected = NullPointerException.class)
    public void nullArgument() {
<span class="nc" id="L34">        Observable.create(null);</span>
<span class="nc" id="L35">    }</span>

    @Test
    public void basic() {
<span class="fc" id="L39">        final Disposable d = Disposables.empty();</span>

<span class="fc" id="L41">        Observable.&lt;Integer&gt;create(new ObservableOnSubscribe&lt;Integer&gt;() {</span>
            @Override
            public void subscribe(ObservableEmitter&lt;Integer&gt; e) throws Exception {
<span class="fc" id="L44">                e.setDisposable(d);</span>

<span class="fc" id="L46">                e.onNext(1);</span>
<span class="fc" id="L47">                e.onNext(2);</span>
<span class="fc" id="L48">                e.onNext(3);</span>
<span class="fc" id="L49">                e.onComplete();</span>
<span class="fc" id="L50">                e.onError(new TestException());</span>
<span class="fc" id="L51">                e.onNext(4);</span>
<span class="fc" id="L52">                e.onError(new TestException());</span>
<span class="fc" id="L53">                e.onComplete();</span>
<span class="fc" id="L54">            }</span>
        })
<span class="fc" id="L56">        .test()</span>
<span class="fc" id="L57">        .assertResult(1, 2, 3);</span>

<span class="fc" id="L59">        assertTrue(d.isDisposed());</span>
<span class="fc" id="L60">    }</span>

    @Test
    public void basicWithCancellable() {
<span class="fc" id="L64">        final Disposable d1 = Disposables.empty();</span>
<span class="fc" id="L65">        final Disposable d2 = Disposables.empty();</span>

<span class="fc" id="L67">        Observable.&lt;Integer&gt;create(new ObservableOnSubscribe&lt;Integer&gt;() {</span>
            @Override
            public void subscribe(ObservableEmitter&lt;Integer&gt; e) throws Exception {
<span class="fc" id="L70">                e.setDisposable(d1);</span>
<span class="fc" id="L71">                e.setCancellable(new Cancellable() {</span>
                    @Override
                    public void cancel() throws Exception {
<span class="fc" id="L74">                        d2.dispose();</span>
<span class="fc" id="L75">                    }</span>
                });

<span class="fc" id="L78">                e.onNext(1);</span>
<span class="fc" id="L79">                e.onNext(2);</span>
<span class="fc" id="L80">                e.onNext(3);</span>
<span class="fc" id="L81">                e.onComplete();</span>
<span class="fc" id="L82">                e.onError(new TestException());</span>
<span class="fc" id="L83">                e.onNext(4);</span>
<span class="fc" id="L84">                e.onError(new TestException());</span>
<span class="fc" id="L85">                e.onComplete();</span>
<span class="fc" id="L86">            }</span>
        })
<span class="fc" id="L88">        .test()</span>
<span class="fc" id="L89">        .assertResult(1, 2, 3);</span>

<span class="fc" id="L91">        assertTrue(d1.isDisposed());</span>
<span class="fc" id="L92">        assertTrue(d2.isDisposed());</span>
<span class="fc" id="L93">    }</span>

    @Test
    public void basicWithError() {
<span class="fc" id="L97">        final Disposable d = Disposables.empty();</span>

<span class="fc" id="L99">        Observable.&lt;Integer&gt;create(new ObservableOnSubscribe&lt;Integer&gt;() {</span>
            @Override
            public void subscribe(ObservableEmitter&lt;Integer&gt; e) throws Exception {
<span class="fc" id="L102">                e.setDisposable(d);</span>

<span class="fc" id="L104">                e.onNext(1);</span>
<span class="fc" id="L105">                e.onNext(2);</span>
<span class="fc" id="L106">                e.onNext(3);</span>
<span class="fc" id="L107">                e.onError(new TestException());</span>
<span class="fc" id="L108">                e.onComplete();</span>
<span class="fc" id="L109">                e.onNext(4);</span>
<span class="fc" id="L110">                e.onError(new TestException());</span>
<span class="fc" id="L111">            }</span>
        })
<span class="fc" id="L113">        .test()</span>
<span class="fc" id="L114">        .assertFailure(TestException.class, 1, 2, 3);</span>

<span class="fc" id="L116">        assertTrue(d.isDisposed());</span>
<span class="fc" id="L117">    }</span>

    @Test
    public void basicSerialized() {
<span class="fc" id="L121">        final Disposable d = Disposables.empty();</span>

<span class="fc" id="L123">        Observable.&lt;Integer&gt;create(new ObservableOnSubscribe&lt;Integer&gt;() {</span>
            @Override
            public void subscribe(ObservableEmitter&lt;Integer&gt; e) throws Exception {
<span class="fc" id="L126">                e = e.serialize();</span>

<span class="fc" id="L128">                e.setDisposable(d);</span>

<span class="fc" id="L130">                e.onNext(1);</span>
<span class="fc" id="L131">                e.onNext(2);</span>
<span class="fc" id="L132">                e.onNext(3);</span>
<span class="fc" id="L133">                e.onComplete();</span>
<span class="fc" id="L134">                e.onError(new TestException());</span>
<span class="fc" id="L135">                e.onNext(4);</span>
<span class="fc" id="L136">                e.onError(new TestException());</span>
<span class="fc" id="L137">                e.onComplete();</span>
<span class="fc" id="L138">            }</span>
        })
<span class="fc" id="L140">        .test()</span>
<span class="fc" id="L141">        .assertResult(1, 2, 3);</span>

<span class="fc" id="L143">        assertTrue(d.isDisposed());</span>
<span class="fc" id="L144">    }</span>

    @Test
    public void basicWithErrorSerialized() {
<span class="fc" id="L148">        final Disposable d = Disposables.empty();</span>

<span class="fc" id="L150">        Observable.&lt;Integer&gt;create(new ObservableOnSubscribe&lt;Integer&gt;() {</span>
            @Override
            public void subscribe(ObservableEmitter&lt;Integer&gt; e) throws Exception {
<span class="fc" id="L153">                e = e.serialize();</span>

<span class="fc" id="L155">                e.setDisposable(d);</span>

<span class="fc" id="L157">                e.onNext(1);</span>
<span class="fc" id="L158">                e.onNext(2);</span>
<span class="fc" id="L159">                e.onNext(3);</span>
<span class="fc" id="L160">                e.onError(new TestException());</span>
<span class="fc" id="L161">                e.onComplete();</span>
<span class="fc" id="L162">                e.onNext(4);</span>
<span class="fc" id="L163">                e.onError(new TestException());</span>
<span class="fc" id="L164">            }</span>
        })
<span class="fc" id="L166">        .test()</span>
<span class="fc" id="L167">        .assertFailure(TestException.class, 1, 2, 3);</span>

<span class="fc" id="L169">        assertTrue(d.isDisposed());</span>
<span class="fc" id="L170">    }</span>

    @Test
    public void wrap() {
<span class="fc" id="L174">        Observable.wrap(new ObservableSource&lt;Integer&gt;() {</span>
            @Override
            public void subscribe(Observer&lt;? super Integer&gt; observer) {
<span class="fc" id="L177">                observer.onSubscribe(Disposables.empty());</span>
<span class="fc" id="L178">                observer.onNext(1);</span>
<span class="fc" id="L179">                observer.onNext(2);</span>
<span class="fc" id="L180">                observer.onNext(3);</span>
<span class="fc" id="L181">                observer.onNext(4);</span>
<span class="fc" id="L182">                observer.onNext(5);</span>
<span class="fc" id="L183">                observer.onComplete();</span>
<span class="fc" id="L184">            }</span>
        })
<span class="fc" id="L186">        .test()</span>
<span class="fc" id="L187">        .assertResult(1, 2, 3, 4, 5);</span>
<span class="fc" id="L188">    }</span>

    @Test
    public void unsafe() {
<span class="fc" id="L192">        Observable.unsafeCreate(new ObservableSource&lt;Integer&gt;() {</span>
            @Override
            public void subscribe(Observer&lt;? super Integer&gt; observer) {
<span class="fc" id="L195">                observer.onSubscribe(Disposables.empty());</span>
<span class="fc" id="L196">                observer.onNext(1);</span>
<span class="fc" id="L197">                observer.onNext(2);</span>
<span class="fc" id="L198">                observer.onNext(3);</span>
<span class="fc" id="L199">                observer.onNext(4);</span>
<span class="fc" id="L200">                observer.onNext(5);</span>
<span class="fc" id="L201">                observer.onComplete();</span>
<span class="fc" id="L202">            }</span>
        })
<span class="fc" id="L204">        .test()</span>
<span class="fc" id="L205">        .assertResult(1, 2, 3, 4, 5);</span>
<span class="fc" id="L206">    }</span>

    @Test(expected = IllegalArgumentException.class)
    public void unsafeWithObservable() {
<span class="nc" id="L210">        Observable.unsafeCreate(Observable.just(1));</span>
<span class="nc" id="L211">    }</span>

    @Test
    public void createNullValue() {
<span class="fc" id="L215">        final Throwable[] error = { null };</span>

<span class="fc" id="L217">        Observable.create(new ObservableOnSubscribe&lt;Integer&gt;() {</span>
            @Override
            public void subscribe(ObservableEmitter&lt;Integer&gt; e) throws Exception {
                try {
<span class="fc" id="L221">                    e.onNext(null);</span>
<span class="fc" id="L222">                    e.onNext(1);</span>
<span class="fc" id="L223">                    e.onError(new TestException());</span>
<span class="fc" id="L224">                    e.onComplete();</span>
<span class="nc" id="L225">                } catch (Throwable ex) {</span>
<span class="nc" id="L226">                    error[0] = ex;</span>
<span class="fc" id="L227">                }</span>
<span class="fc" id="L228">            }</span>
        })
<span class="fc" id="L230">        .test()</span>
<span class="fc" id="L231">        .assertFailure(NullPointerException.class);</span>

<span class="fc" id="L233">        assertNull(error[0]);</span>
<span class="fc" id="L234">    }</span>

    @Test
    public void createNullValueSerialized() {
<span class="fc" id="L238">        final Throwable[] error = { null };</span>

<span class="fc" id="L240">        Observable.create(new ObservableOnSubscribe&lt;Integer&gt;() {</span>
            @Override
            public void subscribe(ObservableEmitter&lt;Integer&gt; e) throws Exception {
<span class="fc" id="L243">                e = e.serialize();</span>
                try {
<span class="fc" id="L245">                    e.onNext(null);</span>
<span class="fc" id="L246">                    e.onNext(1);</span>
<span class="fc" id="L247">                    e.onError(new TestException());</span>
<span class="fc" id="L248">                    e.onComplete();</span>
<span class="nc" id="L249">                } catch (Throwable ex) {</span>
<span class="nc" id="L250">                    error[0] = ex;</span>
<span class="fc" id="L251">                }</span>
<span class="fc" id="L252">            }</span>
        })
<span class="fc" id="L254">        .test()</span>
<span class="fc" id="L255">        .assertFailure(NullPointerException.class);</span>

<span class="fc" id="L257">        assertNull(error[0]);</span>
<span class="fc" id="L258">    }</span>

    @Test
    public void callbackThrows() {
<span class="fc" id="L262">        Observable.create(new ObservableOnSubscribe&lt;Object&gt;() {</span>
            @Override
            public void subscribe(ObservableEmitter&lt;Object&gt; e) throws Exception {
<span class="fc" id="L265">                throw new TestException();</span>
            }
        })
<span class="fc" id="L268">        .test()</span>
<span class="fc" id="L269">        .assertFailure(TestException.class);</span>
<span class="fc" id="L270">    }</span>

    @Test
    public void nullValue() {
<span class="fc" id="L274">        Observable.create(new ObservableOnSubscribe&lt;Object&gt;() {</span>
            @Override
            public void subscribe(ObservableEmitter&lt;Object&gt; e) throws Exception {
<span class="fc" id="L277">                e.onNext(null);</span>
<span class="fc" id="L278">            }</span>
        })
<span class="fc" id="L280">        .test()</span>
<span class="fc" id="L281">        .assertFailure(NullPointerException.class);</span>
<span class="fc" id="L282">    }</span>

    @Test
    public void nullThrowable() {
<span class="fc" id="L286">        Observable.create(new ObservableOnSubscribe&lt;Object&gt;() {</span>
            @Override
            public void subscribe(ObservableEmitter&lt;Object&gt; e) throws Exception {
<span class="fc" id="L289">                e.onError(null);</span>
<span class="fc" id="L290">            }</span>
        })
<span class="fc" id="L292">        .test()</span>
<span class="fc" id="L293">        .assertFailure(NullPointerException.class);</span>
<span class="fc" id="L294">    }</span>

    @Test
    public void nullValueSync() {
<span class="fc" id="L298">        Observable.create(new ObservableOnSubscribe&lt;Object&gt;() {</span>
            @Override
            public void subscribe(ObservableEmitter&lt;Object&gt; e) throws Exception {
<span class="fc" id="L301">                e.serialize().onNext(null);</span>
<span class="fc" id="L302">            }</span>
        })
<span class="fc" id="L304">        .test()</span>
<span class="fc" id="L305">        .assertFailure(NullPointerException.class);</span>
<span class="fc" id="L306">    }</span>

    @Test
    public void nullThrowableSync() {
<span class="fc" id="L310">        Observable.create(new ObservableOnSubscribe&lt;Object&gt;() {</span>
            @Override
            public void subscribe(ObservableEmitter&lt;Object&gt; e) throws Exception {
<span class="fc" id="L313">                e.serialize().onError(null);</span>
<span class="fc" id="L314">            }</span>
        })
<span class="fc" id="L316">        .test()</span>
<span class="fc" id="L317">        .assertFailure(NullPointerException.class);</span>
<span class="fc" id="L318">    }</span>

    @Test
    public void onErrorCrash() {
<span class="fc" id="L322">        Observable.create(new ObservableOnSubscribe&lt;Object&gt;() {</span>
            @Override
            public void subscribe(ObservableEmitter&lt;Object&gt; e) throws Exception {
<span class="fc" id="L325">                Disposable d = Disposables.empty();</span>
<span class="fc" id="L326">                e.setDisposable(d);</span>
                try {
<span class="nc" id="L328">                    e.onError(new IOException());</span>
<span class="nc" id="L329">                    fail(&quot;Should have thrown&quot;);</span>
<span class="fc" id="L330">                } catch (TestException ex) {</span>
                    // expected
<span class="nc" id="L332">                }</span>
<span class="fc" id="L333">                assertTrue(d.isDisposed());</span>
<span class="fc" id="L334">            }</span>
        })
<span class="fc" id="L336">        .subscribe(new Observer&lt;Object&gt;() {</span>
            @Override
            public void onSubscribe(Disposable d) {
<span class="fc" id="L339">            }</span>

            @Override
            public void onNext(Object value) {
<span class="nc" id="L343">            }</span>

            @Override
            public void onError(Throwable e) {
<span class="fc" id="L347">                throw new TestException();</span>
            }

            @Override
            public void onComplete() {
<span class="nc" id="L352">            }</span>
        });
<span class="fc" id="L354">    }</span>

    @Test
    public void onCompleteCrash() {
<span class="fc" id="L358">        Observable.create(new ObservableOnSubscribe&lt;Object&gt;() {</span>
            @Override
            public void subscribe(ObservableEmitter&lt;Object&gt; e) throws Exception {
<span class="fc" id="L361">                Disposable d = Disposables.empty();</span>
<span class="fc" id="L362">                e.setDisposable(d);</span>
                try {
<span class="nc" id="L364">                    e.onComplete();</span>
<span class="nc" id="L365">                    fail(&quot;Should have thrown&quot;);</span>
<span class="fc" id="L366">                } catch (TestException ex) {</span>
                    // expected
<span class="nc" id="L368">                }</span>
<span class="fc" id="L369">                assertTrue(d.isDisposed());</span>
<span class="fc" id="L370">            }</span>
        })
<span class="fc" id="L372">        .subscribe(new Observer&lt;Object&gt;() {</span>
            @Override
            public void onSubscribe(Disposable d) {
<span class="fc" id="L375">            }</span>

            @Override
            public void onNext(Object value) {
<span class="nc" id="L379">            }</span>

            @Override
            public void onError(Throwable e) {
<span class="nc" id="L383">            }</span>

            @Override
            public void onComplete() {
<span class="fc" id="L387">                throw new TestException();</span>
            }
        });
<span class="fc" id="L390">    }</span>

    @Test
    public void serialized() {
<span class="fc" id="L394">        List&lt;Throwable&gt; errors = TestHelper.trackPluginErrors();</span>
        try {
<span class="fc" id="L396">            Observable.create(new ObservableOnSubscribe&lt;Object&gt;() {</span>
                @Override
                public void subscribe(ObservableEmitter&lt;Object&gt; e) throws Exception {
<span class="fc" id="L399">                    ObservableEmitter&lt;Object&gt; f = e.serialize();</span>

<span class="fc" id="L401">                    assertSame(f, f.serialize());</span>

<span class="fc" id="L403">                    assertFalse(f.isDisposed());</span>

<span class="fc" id="L405">                    final int[] calls = { 0 };</span>

<span class="fc" id="L407">                    f.setCancellable(new Cancellable() {</span>
                        @Override
                        public void cancel() throws Exception {
<span class="fc" id="L410">                            calls[0]++;</span>
<span class="fc" id="L411">                        }</span>
                    });

<span class="fc" id="L414">                    e.onComplete();</span>

<span class="fc" id="L416">                    assertTrue(f.isDisposed());</span>

<span class="fc" id="L418">                    assertEquals(1, calls[0]);</span>
<span class="fc" id="L419">                }</span>
            })
<span class="fc" id="L421">            .test()</span>
<span class="fc" id="L422">            .assertResult();</span>

<span class="fc" id="L424">            assertTrue(errors.toString(), errors.isEmpty());</span>
        } finally {
<span class="fc" id="L426">            RxJavaPlugins.reset();</span>
        }
<span class="fc" id="L428">    }</span>

    @Test
    public void serializedConcurrentOnNext() {
<span class="fc" id="L432">        Observable.create(new ObservableOnSubscribe&lt;Object&gt;() {</span>
            @Override
            public void subscribe(ObservableEmitter&lt;Object&gt; e) throws Exception {
<span class="fc" id="L435">                final ObservableEmitter&lt;Object&gt; f = e.serialize();</span>

<span class="fc" id="L437">                Runnable r1 = new Runnable() {</span>
                    @Override
                    public void run() {
<span class="fc bfc" id="L440" title="All 2 branches covered.">                        for (int i = 0; i &lt; TestHelper.RACE_DEFAULT_LOOPS; i++) {</span>
<span class="fc" id="L441">                            f.onNext(1);</span>
                        }
<span class="fc" id="L443">                    }</span>
                };

<span class="fc" id="L446">                TestHelper.race(r1, r1);</span>
<span class="fc" id="L447">            }</span>
        })
<span class="fc" id="L449">        .take(TestHelper.RACE_DEFAULT_LOOPS)</span>
<span class="fc" id="L450">        .test()</span>
<span class="fc" id="L451">        .assertSubscribed().assertValueCount(TestHelper.RACE_DEFAULT_LOOPS).assertComplete().assertNoErrors();</span>
<span class="fc" id="L452">    }</span>

    @Test
    public void serializedConcurrentOnNextOnError() {
<span class="fc" id="L456">        Observable.create(new ObservableOnSubscribe&lt;Object&gt;() {</span>
            @Override
            public void subscribe(ObservableEmitter&lt;Object&gt; e) throws Exception {
<span class="fc" id="L459">                final ObservableEmitter&lt;Object&gt; f = e.serialize();</span>

<span class="fc" id="L461">                Runnable r1 = new Runnable() {</span>
                    @Override
                    public void run() {
<span class="fc bfc" id="L464" title="All 2 branches covered.">                        for (int i = 0; i &lt; 1000; i++) {</span>
<span class="fc" id="L465">                            f.onNext(1);</span>
                        }
<span class="fc" id="L467">                    }</span>
                };

<span class="fc" id="L470">                Runnable r2 = new Runnable() {</span>
                    @Override
                    public void run() {
<span class="fc bfc" id="L473" title="All 2 branches covered.">                        for (int i = 0; i &lt; 100; i++) {</span>
<span class="fc" id="L474">                            f.onNext(1);</span>
                        }
<span class="fc" id="L476">                        f.onError(new TestException());</span>
<span class="fc" id="L477">                    }</span>
                };

<span class="fc" id="L480">                TestHelper.race(r1, r2);</span>
<span class="fc" id="L481">            }</span>
        })
<span class="fc" id="L483">        .test()</span>
<span class="fc" id="L484">        .assertSubscribed().assertNotComplete()</span>
<span class="fc" id="L485">        .assertError(TestException.class);</span>
<span class="fc" id="L486">    }</span>

    @Test
    public void serializedConcurrentOnNextOnComplete() {
<span class="fc" id="L490">        TestObserver&lt;Object&gt; to = Observable.create(new ObservableOnSubscribe&lt;Object&gt;() {</span>
            @Override
            public void subscribe(ObservableEmitter&lt;Object&gt; e) throws Exception {
<span class="fc" id="L493">                final ObservableEmitter&lt;Object&gt; f = e.serialize();</span>

<span class="fc" id="L495">                Runnable r1 = new Runnable() {</span>
                    @Override
                    public void run() {
<span class="fc bfc" id="L498" title="All 2 branches covered.">                        for (int i = 0; i &lt; 1000; i++) {</span>
<span class="fc" id="L499">                            f.onNext(1);</span>
                        }
<span class="fc" id="L501">                    }</span>
                };

<span class="fc" id="L504">                Runnable r2 = new Runnable() {</span>
                    @Override
                    public void run() {
<span class="fc bfc" id="L507" title="All 2 branches covered.">                        for (int i = 0; i &lt; 100; i++) {</span>
<span class="fc" id="L508">                            f.onNext(1);</span>
                        }
<span class="fc" id="L510">                        f.onComplete();</span>
<span class="fc" id="L511">                    }</span>
                };

<span class="fc" id="L514">                TestHelper.race(r1, r2);</span>
<span class="fc" id="L515">            }</span>
        })
<span class="fc" id="L517">        .test()</span>
<span class="fc" id="L518">        .assertSubscribed().assertComplete()</span>
<span class="fc" id="L519">        .assertNoErrors();</span>

<span class="fc" id="L521">        int c = to.valueCount();</span>
<span class="pc bpc" id="L522" title="1 of 2 branches missed.">        assertTrue(&quot;&quot; + c, c &gt;= 100);</span>
<span class="fc" id="L523">    }</span>

    @Test
    public void onErrorRace() {
<span class="fc" id="L527">        Observable&lt;Object&gt; source = Observable.create(new ObservableOnSubscribe&lt;Object&gt;() {</span>
            @Override
            public void subscribe(ObservableEmitter&lt;Object&gt; e) throws Exception {
<span class="fc" id="L530">                final ObservableEmitter&lt;Object&gt; f = e.serialize();</span>

<span class="fc" id="L532">                final TestException ex = new TestException();</span>

<span class="fc" id="L534">                Runnable r1 = new Runnable() {</span>
                    @Override
                    public void run() {
<span class="fc" id="L537">                        f.onError(null);</span>
<span class="fc" id="L538">                    }</span>
                };

<span class="fc" id="L541">                Runnable r2 = new Runnable() {</span>
                    @Override
                    public void run() {
<span class="fc" id="L544">                        f.onError(ex);</span>
<span class="fc" id="L545">                    }</span>
                };

<span class="fc" id="L548">                TestHelper.race(r1, r2);</span>
<span class="fc" id="L549">            }</span>
        });

<span class="fc" id="L552">        List&lt;Throwable&gt; errors = TestHelper.trackPluginErrors();</span>

        try {
<span class="fc bfc" id="L555" title="All 2 branches covered.">            for (int i = 0; i &lt; TestHelper.RACE_DEFAULT_LOOPS; i++) {</span>
<span class="fc" id="L556">                source</span>
<span class="fc" id="L557">                .test()</span>
<span class="fc" id="L558">                .assertFailure(Throwable.class);</span>
            }
        } finally {
<span class="fc" id="L561">            RxJavaPlugins.reset();</span>
        }
<span class="fc" id="L563">        assertFalse(errors.isEmpty());</span>
<span class="fc" id="L564">    }</span>

    @Test
    public void onCompleteRace() {
<span class="fc" id="L568">        Observable&lt;Object&gt; source = Observable.create(new ObservableOnSubscribe&lt;Object&gt;() {</span>
            @Override
            public void subscribe(ObservableEmitter&lt;Object&gt; e) throws Exception {
<span class="fc" id="L571">                final ObservableEmitter&lt;Object&gt; f = e.serialize();</span>

<span class="fc" id="L573">                Runnable r1 = new Runnable() {</span>
                    @Override
                    public void run() {
<span class="fc" id="L576">                        f.onComplete();</span>
<span class="fc" id="L577">                    }</span>
                };

<span class="fc" id="L580">                Runnable r2 = new Runnable() {</span>
                    @Override
                    public void run() {
<span class="fc" id="L583">                        f.onComplete();</span>
<span class="fc" id="L584">                    }</span>
                };

<span class="fc" id="L587">                TestHelper.race(r1, r2);</span>
<span class="fc" id="L588">            }</span>
        });

<span class="fc bfc" id="L591" title="All 2 branches covered.">        for (int i = 0; i &lt; TestHelper.RACE_DEFAULT_LOOPS; i++) {</span>
<span class="fc" id="L592">            source</span>
<span class="fc" id="L593">            .test()</span>
<span class="fc" id="L594">            .assertResult();</span>
        }
<span class="fc" id="L596">    }</span>

    @Test
    public void tryOnError() {
<span class="fc" id="L600">        List&lt;Throwable&gt; errors = TestHelper.trackPluginErrors();</span>
        try {
<span class="fc" id="L602">            final Boolean[] response = { null };</span>
<span class="fc" id="L603">            Observable.create(new ObservableOnSubscribe&lt;Object&gt;() {</span>
                @Override
                public void subscribe(ObservableEmitter&lt;Object&gt; e) throws Exception {
<span class="fc" id="L606">                    e.onNext(1);</span>
<span class="fc" id="L607">                    response[0] = e.tryOnError(new TestException());</span>
<span class="fc" id="L608">                }</span>
            })
<span class="fc" id="L610">            .take(1)</span>
<span class="fc" id="L611">            .test()</span>
<span class="fc" id="L612">            .assertResult(1);</span>

<span class="fc" id="L614">            assertFalse(response[0]);</span>

<span class="fc" id="L616">            assertTrue(errors.toString(), errors.isEmpty());</span>
        } finally {
<span class="fc" id="L618">            RxJavaPlugins.reset();</span>
        }
<span class="fc" id="L620">    }</span>

    @Test
    public void tryOnErrorSerialized() {
<span class="fc" id="L624">        List&lt;Throwable&gt; errors = TestHelper.trackPluginErrors();</span>
        try {
<span class="fc" id="L626">            final Boolean[] response = { null };</span>
<span class="fc" id="L627">            Observable.create(new ObservableOnSubscribe&lt;Object&gt;() {</span>
                @Override
                public void subscribe(ObservableEmitter&lt;Object&gt; e) throws Exception {
<span class="fc" id="L630">                    e = e.serialize();</span>
<span class="fc" id="L631">                    e.onNext(1);</span>
<span class="fc" id="L632">                    response[0] = e.tryOnError(new TestException());</span>
<span class="fc" id="L633">                }</span>
            })
<span class="fc" id="L635">            .take(1)</span>
<span class="fc" id="L636">            .test()</span>
<span class="fc" id="L637">            .assertResult(1);</span>

<span class="fc" id="L639">            assertFalse(response[0]);</span>

<span class="fc" id="L641">            assertTrue(errors.toString(), errors.isEmpty());</span>
        } finally {
<span class="fc" id="L643">            RxJavaPlugins.reset();</span>
        }
<span class="fc" id="L645">    }</span>

    @Test
    public void emitterHasToString() {
<span class="fc" id="L649">        Observable.create(new ObservableOnSubscribe&lt;Object&gt;() {</span>
            @Override
            public void subscribe(ObservableEmitter&lt;Object&gt; emitter) throws Exception {
<span class="fc" id="L652">                assertTrue(emitter.toString().contains(ObservableCreate.CreateEmitter.class.getSimpleName()));</span>
<span class="fc" id="L653">                assertTrue(emitter.serialize().toString().contains(ObservableCreate.CreateEmitter.class.getSimpleName()));</span>
<span class="fc" id="L654">            }</span>
<span class="fc" id="L655">        }).test().assertEmpty();</span>
<span class="fc" id="L656">    }</span>
}
</pre><div class="footer"><span class="right">Created with <a href="http://www.jacoco.org/jacoco">JaCoCo</a> 0.8.2.201808211720</span></div></body></html>